test/integration/thruster.cc

```diff
@@ -0,0 +1,216 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <gtest/gtest.h>
+
+#include <ignition/msgs/double.pb.h>
+
+#include <ignition/common/Console.hh>
+#include <ignition/common/Util.hh>
+#include <ignition/transport/Node.hh>
+
+#include "ignition/gazebo/Link.hh"
+#include "ignition/gazebo/Model.hh"
+#include "ignition/gazebo/Server.hh"
+#include "ignition/gazebo/SystemLoader.hh"
+#include "ignition/gazebo/TestFixture.hh"
+#include "ignition/gazebo/Util.hh"
+#include "ignition/gazebo/World.hh"
+
+#include "ignition/gazebo/test_config.hh"
+#include "../helpers/EnvTestFixture.hh"
+
+using namespace ignition;
+using namespace gazebo;
+
+class ThrusterTest : public InternalFixture<::testing::Test>
+{
+  /// \brief Test a world file
+  /// \param[in] _world Path to world file
+  /// \param[in] _namespace Namespace for topic
+  /// \param[in] _coefficient Thrust coefficient
+  /// \param[in] _density Fluid density
+  /// \param[in] _diameter Propeller diameter
+  /// \param[in] _baseTol Base tolerance for most quantities
+  public: void TestWorld(const std::string &_world,
+      const std::string &_namespace, double _coefficient, double _density,
+      double _diameter, double _baseTol);
+};
+
+//////////////////////////////////////////////////
+void ThrusterTest::TestWorld(const std::string &_world,
+    const std::string &_namespace, double _coefficient, double _density,
+    double _diameter, double _baseTol)
+{
+  // Start server
+  ServerConfig serverConfig;
+  serverConfig.SetSdfFile(_world);
+
+  TestFixture fixture(serverConfig);
+
+  Model model;
+  Link propeller;
+  std::vector<math::Pose3d> modelPoses;
+  std::vector<math::Vector3d> propellerAngVels;
+  double dt{0.0};
+  fixture.
+  OnConfigure(
+    [&](const ignition::gazebo::Entity &_worldEntity,
+      const std::shared_ptr<const sdf::Element> &/*_sdf*/,
+      ignition::gazebo::EntityComponentManager &_ecm,
+      ignition::gazebo::EventManager &/*_eventMgr*/)
+    {
+      World world(_worldEntity);
+
+      auto modelEntity = world.ModelByName(_ecm, "sub");
+      EXPECT_NE(modelEntity, kNullEntity);
+      model = Model(modelEntity);
+
+      auto propellerEntity = model.LinkByName(_ecm, "propeller");
+      EXPECT_NE(propellerEntity, kNullEntity);
+
+      propeller = Link(propellerEntity);
+      propeller.EnableVelocityChecks(_ecm);
+    }).
+  OnPostUpdate([&](const gazebo::UpdateInfo &_info,
+                            const gazebo::EntityComponentManager &_ecm)
+    {
+      dt = std::chrono::duration_cast<std::chrono::nanoseconds>(
+          _info.dt).count() * 1e-9;
```

Comment:
Knit: could be done with `std::duration_cast<std::chrono::seconds, double>(_info.dt).count()` thus avoiding the use of 1e-9.

Reply:
Good call 9a0642b